distmv/studentst.go

```diff
@@ -0,0 +1,335 @@
+	dim := len(mu)
+	ob := len(observed)
+
+	unobserved := findUnob(observed, dim)
+
+	unob := len(unobserved)
+	if unob == 0 {
+		panic("stat: all dimensions observed")
+	}
+
+	mu1 := make([]float64, unob)
+	for i, v := range unobserved {
+		mu1[i] = mu[v]
+	}
+	mu2 := make([]float64, ob) // really v - mu2
+	for i, v := range observed {
+		mu2[i] = values[i] - mu[v]
+	}
+
+	var sigma11, sigma22 mat64.SymDense
+	sigma11.SubsetSym(sigma, unobserved)
+	sigma22.SubsetSym(sigma, observed)
+
+	sigma21 := mat64.NewDense(ob, unob, nil)
+	for i, r := range observed {
+		for j, c := range unobserved {
+			v := sigma.At(r, c)
+			sigma21.Set(i, j, v)
+		}
+	}
+
+	var chol mat64.Cholesky
+	ok := chol.Factorize(&sigma22)
+	if !ok {
+		return math.NaN(), nil, nil
+	}
+
+	// Compute mu_1 + sigma_{2,1}^T * sigma_{2,2}^-1 (v - mu_2).
+	v := mat64.NewVector(ob, mu2)
+	var tmp, tmp2 mat64.Vector
+	err := tmp.SolveCholeskyVec(&chol, v)
+	if err != nil {
+		return math.NaN(), nil, nil
+	}
+	tmp2.MulVec(sigma21.T(), &tmp)
+
+	for i := range mu1 {
+		mu1[i] += tmp2.At(i, 0)
+	}
+
+	// Compute sigma_{1,1} - sigma_{2,1}^T * sigma_{2,2}^-1 * sigma_{2,1}.
+	// TODO(btracey): Should this be a method of SymDense?
+	var tmp3, tmp4 mat64.Dense
+	err = tmp3.SolveCholesky(&chol, sigma21)
+	if err != nil {
+		return math.NaN(), nil, nil
+	}
+	tmp4.Mul(sigma21.T(), &tmp3)
+
+	// TODO(btracey): If tmp4 can constructed with a method, then this can be
+	// replaced with SubSym.
+	for i := 0; i < len(unobserved); i++ {
+		for j := i; j < len(unobserved); j++ {
+			v := sigma11.At(i, j)
+			sigma11.SetSym(i, j, v-tmp4.At(i, j))
+		}
+	}
+
+	// The computed variables are accurate for a Normal.
+	if math.IsInf(nu, 1) {
+		return nu, mu1, &sigma11
+	}
+
+	// Compute beta = (v - mu_2)^T * sigma_{2,2}^-1 * (v - mu_2)^T
+	beta := mat64.Dot(v, &tmp)
+
+	// Scale the covariance matrix
+	sigma11.ScaleSym((nu+beta)/(nu+float64(ob)), &sigma11)
+
+	return nu + float64(ob), mu1, &sigma11
+}
+
+// CovarianceMatrix returns the covariance matrix of the distribution. Upon
+// return, the value at element {i, j} of the covariance matrix is equal to
+// the covariance of the i^th and j^th variables.
+//  covariance(i, j) = E[(x_i - E[x_i])(x_j - E[x_j])]
+// If the input matrix is nil a new matrix is allocated, otherwise the result
+// is stored in-place into the input.
+func (st *StudentsT) CovarianceMatrix(s *mat64.SymDense) *mat64.SymDense {
+	if s == nil {
+		s = mat64.NewSymDense(st.dim, nil)
+	}
+	sn := s.Symmetric()
+	if sn != st.dim {
+		panic("normal: input matrix size mismatch")
+	}
+	st.setSigma()
+	s.CopySym(st.sigma)
+	s.ScaleSym(st.nu/(st.nu-2), s)
+	return s
+}
+
+// Dim returns the dimension of the distribution.
+func (s *StudentsT) Dim() int {
+	return s.dim
+}
+
+func (s *StudentsT) LogProb(y []float64) float64 {
+	if len(y) != s.dim {
+		panic(badInputLength)
+	}
+
+	nu := s.nu
```

Comment:
It's just the formula in the wikipedia entry, or in the pdf linked in the struct documentation. What kind of commentary would you like?